packages/protocol/contracts/bridge/libs/LibBridgeRelease.sol

```diff
@@ -56,37 +60,99 @@ library LibBridgeRelease {
 
         bytes32 msgHash = message.hashMessage();
 
-        if (state.etherReleased[msgHash] == true) {
-            revert B_ETHER_RELEASED_ALREADY();
-        }
-
         if (
             !LibBridgeStatus.isMessageFailed(
                 resolver, msgHash, message.destChainId, proof
             )
         ) {
             revert B_MSG_NOT_FAILED();
         }
+        
+        if(state.recallStatus[msgHash] 
+                == LibBridgeData.RecallStatus.ETH_AND_TOKEN_RELEASED
+        ){
+            // Both ether and tokens are released
+            revert B_TOKENS_RELEASED_ALREADY();
+        }
+
+        uint256 releaseAmount;
 
-        state.etherReleased[msgHash] = true;
+        if(state.recallStatus[msgHash] 
+                == LibBridgeData.RecallStatus.NOT_RECALLED
+        ) {
+            // Release ETH first
+            state.recallStatus[msgHash] = LibBridgeData.RecallStatus.ETH_RELEASED;
 
-        uint256 releaseAmount = message.depositValue + message.callValue;
+            releaseAmount = message.depositValue + message.callValue;
 
-        if (releaseAmount > 0) {
-            address ethVault = resolver.resolve("ether_vault", true);
-            // if on Taiko
-            if (ethVault != address(0)) {
-                EtherVault(payable(ethVault)).releaseEther(
-                    message.owner, releaseAmount
-                );
-            } else {
-                // if on Ethereum
-                (bool success,) = message.owner.call{ value: releaseAmount }("");
-                if (!success) {
-                    revert B_FAILED_TRANSFER();
+            if (releaseAmount > 0) {
+                address ethVault = resolver.resolve("ether_vault", true);
+                // if on Taiko
+                if (ethVault != address(0)) {
+                    EtherVault(payable(ethVault)).releaseEther(
+                        message.owner, releaseAmount
+                    );
+                } else {
+                    // if on Ethereum
+                    (bool success,) = message.owner.call{ value: releaseAmount }("");
+                    if (!success) {
+                        revert B_FAILED_TRANSFER();
+                    }
+                }
+            }
+        }
+        //2nd stage is releasing the tokens
+        if(state.recallStatus[msgHash] 
+                == LibBridgeData.RecallStatus.ETH_RELEASED 
+                && message.to != address(0)
+                && message.data.length != 0
+        ) {
+            // We now, need to know which tokenVault 'IS' the one. So from message.to - we cannot
```

Comment:
In all bridge contract, if you seach for "Vault" and find it, you should remove it. Bridge is designed to be generate purpose, it's not a vault master or something. `to` can be any address written and deployed by others.